src/dattobd.c

```diff
@@ -4121,13 +4121,95 @@ static int ioctl_dattobd_info(struct dattobd_info *info){
 	return ret;
 }
 
+static int ioctl_dattobd_active_device_info(struct dattobd_active_device_info **adinfo, unsigned int *copyback)
+{
+    /* get the count from *info, free the memory, figure out how much we're actually going to return
+     * reallocate *info to that size, then copy the structs in, set the count for what we're returning
+     * and we're done. */
+	int ret;
+	int active;
+	int user_max; // the count of how many structs the caller has reserved space for
+	int lp;
+	int pos;
+    unsigned long bufsz;
+    struct snap_device *dev;
+    struct dattobd_info *infoitem;
+
+	user_max = (*adinfo)->count;
+	LOG_DEBUG("received dattobd all device info ioctl - max return count %d", user_max);
+
+	// count how many devices there are to return
+	active = 0;
+	for (lp = 0; lp < dattobd_max_snap_devices; lp++)
```

Comment:
Actually, I would use either the macro `tracer_for_each()` or `tracer_for_each_full()`. These functions correctly interface with `snap_devices` as an RCU array. This is not strictly necessary here, under the ioctl lock, but it keeps the code cleaner if we use the same code for looping through devices.